File: test/TEST_FrameProtocol.cs
using System;
using System.Collections;
using System.Collections.Generic;
using System.Text;

using Xunit;

using InoIPC;

namespace InoIPC.TEST
{
   // ============================================================
   /// <summary>
   /// Tests for FrameProtocol round-trip via TestTransport.
   /// </summary>
   // ============================================================
   public class TEST_FrameProtocol
   {

   #region Round-trip

      [Fact]
      public void SendReceive_RoundTrip()
      {
         var transport = new TestTransport();

         FrameProtocol.Send(transport, "{\"test\":\"ping\"}");

         string result = FrameProtocol.Receive(transport);

         Assert.Equal("{\"test\":\"ping\"}", result);
      }

      [Fact]
      public void SendReceive_EmptyBody()
      {
         var transport = new TestTransport();

         FrameProtocol.Send(transport, "");

         string result = FrameProtocol.Receive(transport);

         Assert.Equal("", result);
      }

      [Fact]
      public void SendReceive_Unicode()
      {
         var transport = new TestTransport();
         string input  = "{\"msg\":\"한글 테스트 🎉\"}";

         FrameProtocol.Send(transport, input);

         string result = FrameProtocol.Receive(transport);

         Assert.Equal(input, result);
      }

      [Fact]
      public void SendReceive_LargePayload()
      {
         var transport = new TestTransport();
         string input  = new string('x', 100_000);

         FrameProtocol.Send(transport, input);

         string result = FrameProtocol.Receive(transport);

         Assert.Equal(input, result);
      }

      [Fact]
      public void SendReceive_MultipleFrames()
      {
         var transport = new TestTransport();

         FrameProtocol.Send(transport, "first");
         FrameProtocol.Send(transport, "second");
         FrameProtocol.Send(transport, "third");

         Assert.Equal("first", FrameProtocol.Receive(transport));
         Assert.Equal("second", FrameProtocol.Receive(transport));
         Assert.Equal("third", FrameProtocol.Receive(transport));
      }

   #endregion

   }
}

File: rust/inoipc/tests/test_frame_protocol.rs
use inoipc::{TestTransport, frame_receive, frame_send};

#[test]
fn send_receive_round_trip() {
    let mut t = TestTransport::new();
    frame_send(&mut t, "{\"test\":\"ping\"}").unwrap();
    let result = frame_receive(&mut t).unwrap();
    assert_eq!(result, "{\"test\":\"ping\"}");
}

#[test]
fn send_receive_empty_body() {
    let mut t = TestTransport::new();
    frame_send(&mut t, "").unwrap();
    let result = frame_receive(&mut t).unwrap();
    assert_eq!(result, "");
}

#[test]
fn send_receive_unicode() {
    let mut t = TestTransport::new();
    let input = "{\"msg\":\"한글 테스트 🎉\"}";
    frame_send(&mut t, input).unwrap();
    let result = frame_receive(&mut t).unwrap();
    assert_eq!(result, input);
}

#[test]
fn send_receive_large_payload() {
    let mut t = TestTransport::new();
    let input: String = "x".repeat(100_000);
    frame_send(&mut t, &input).unwrap();
    let result = frame_receive(&mut t).unwrap();
    assert_eq!(result, input);
}

#[test]
fn send_receive_multiple_frames() {
    let mut t = TestTransport::new();
    frame_send(&mut t, "first").unwrap();
    frame_send(&mut t, "second").unwrap();
    frame_send(&mut t, "third").unwrap();
    assert_eq!(frame_receive(&mut t).unwrap(), "first");
    assert_eq!(frame_receive(&mut t).unwrap(), "second");
    assert_eq!(frame_receive(&mut t).unwrap(), "third");
}
